### PR TITLE
Add native OIDC support in Grist #707

### LINKED ISSUE #707: Support OIDC directly
It would simplify our architecture if Grist could support OIDC directly, just like it supports SAML.

There seem to be many people also wishing to use OIDC in this issue: https://github.com/gristlabs/grist-core/issues/44

What do you think?

### PR DESCRIPTION
**Pull request by fflorent**
Resolves #707

# Proposed implementation

We use the `openid-client` package, which is certified as [openid compliant](https://www.npmjs.com/package/openid-client#certification).
Also the code is heavily inspired from [SamlConfig](https://github.com/gristlabs/grist-core/blob/d29e23efe83d71f0e3daf921898cac2efcdce8cc/app/server/lib/SamlConfig.ts)

### CODE REVIEW

File: app/server/lib/OIDCConfig.ts
Comment (fflorent) [inline app/server/lib/OIDCConfig.ts:51]: I am pretty sure this is not the right way to initialize a nonce. Instead it should be instantiated each time a user attempts to login.

Probably needs a way to store it. Where?

File: app/server/lib/OIDCConfig.ts
Comment (hihuz) [inline app/server/lib/OIDCConfig.ts:3]: Nit: SAML -> OpenID Connect

```suggestion
 * A good informative overview of OIDC is at https://developer.okta.com/blog/2019/10/21/illustrated-guide-to-oauth-and-oidc
```

File: app/server/lib/OIDCConfig.ts
Comment (hihuz) [inline app/server/lib/OIDCConfig.ts:64]: Nit: probably should be `initOIDC` instead of `initConfig` in these error messages?

**Review by hihuz — COMMENTED** [review]
Looks clean and quite consistent with other login configurations 👍 

I agree with your comment, although I am not familiar enough with the grist codebase to give a good advice here 🙁

File: app/server/lib/OIDCConfig.ts
Comment (fflorent) [reply app/server/lib/OIDCConfig.ts:51]: So instead of using `generators.codeVerifier()` included in node-openid-client, which [generates random bytes and stringifies it](https://github.com/panva/node-openid-client/blob/a84d022f195f82ca1c97f8f6b2567ebcef8738c3/lib/helpers/generators.js#L5C32-L5C41), we could take advantage of the permit store, which generates some random string taking advantage of the UUIDv4 algorithm and store it temporarily? I give it a try

PS: I think using the code_verifier alone is enough, neither the `nonce` nor the `state` are useful when dealing with the authorization code flow: https://github.com/panva/node-openid-client/tree/main#authorization-code-flow

Comment (fflorent) [reply app/server/lib/OIDCConfig.ts]: I tried something, still a bit worried about the implementation regarding the security (the codeVerifier is stored in the permit store the key returned is passed to the identity provider as a GET parameter (as the state), which worries me a bit even if I have no clear idea how an attacker could operate).

File: app/server/lib/OIDCConfig.ts
Comment (dsagal) [inline app/server/lib/OIDCConfig.ts:74]: Is this endpoint (`LOGIN_URL`) needed at all? I think `/login`, `/signup`, and `/signin` might all do the same thing (by redirecting ultimately to `this.getLoginRedirectUrl()`)

File: app/server/lib/OIDCConfig.ts
Comment (dsagal) [inline app/server/lib/OIDCConfig.ts:88]: I think it would be better to store this URL as a member variable of OIDCConfig, since it's used in a couple of places (also in `initOIDC`) -- it would make it clearer that it's intended to be the same value.

File: app/server/lib/OIDCConfig.ts
Comment (dsagal) [inline app/server/lib/OIDCConfig.ts:97]: I think these tokens (`accessToken` and `refreshToken`) aren't ever used. I.e. `accessToken` is used to get profile info, and that's it. Any reason to store them in the session? One reason would be to verify periodically that the user is still active with the issuer, but I don't think this implementation (or the SAML one) currently do that.

File: app/server/lib/OIDCConfig.ts
Comment (dsagal) [inline app/server/lib/OIDCConfig.ts:103]: I wonder if this message may be misleading -- it seems there could be other reasons for failure than failure to connect (e.g. expired permit)

File: app/server/lib/OIDCConfig.ts
Comment (dsagal) [inline app/server/lib/OIDCConfig.ts:94]: In the SAML implementation, there is an extra argument `{sessionId: state.sessionId}`, with the explanation that the cookie may not be included in the callback request (when it's redirected from the issuer). Maybe this isn't needed here because the callback is a GET request (rather than POST), but just wanted to confirm.

File: app/server/lib/OIDCConfig.ts
Comment (dsagal) [inline app/server/lib/OIDCConfig.ts:45]: Maybe add a comment that this is used in temporary permits?

**Review by dsagal — COMMENTED** [review]
Hello @fflorent ! I am jumping in with a review, since I did much of the SAML implementation. Overall it looks great, and most of my comments are about clarifying things.

File: app/server/lib/OIDCConfig.ts
Comment (fflorent) [reply app/server/lib/OIDCConfig.ts:29]: I saw this paragraph much as a way to test the authentication flow locally, typically when working on this module, not for production purpose. My concern is that a developer can easily set up a local environment to start working.

I can add a warning so the reader would know how important it is to use a secure connection. Would that be enough?

File: app/server/lib/OIDCConfig.ts
Comment (fflorent) [reply app/server/lib/OIDCConfig.ts:74]: You're right! It's useless, I am removing it

File: app/server/lib/OIDCConfig.ts
Comment (fflorent) [reply app/server/lib/OIDCConfig.ts:97]: No true reason, indeed. If we need that in the future, we would know that's easy to add this info.

File: app/server/lib/OIDCConfig.ts
Comment (fflorent) [reply app/server/lib/OIDCConfig.ts:94]: Right. In my case, extracting the sessionId worked well using the cookie, but after your message, I am a bit unsure.

I was about to add a check to ensure that the sessionId matches between the permit and the cookie.

However, if we cannot rely on the cookie, if I am correct, that would mean that the code challenge can be retrieved only thanks to the value of the `state` variable, which means that the code challenge is not much useful here (protection offered by the PKCE is explained here: <https://dzone.com/articles/what-is-pkce>)

I am looking for a way to store the code verifier in the session, but it appears that at the contrary of some other session mechanisms, I have to retrieve some randomly generated key and store it by myself. That's the reason why I am passing this random key as the state, but I am not very confident of that way of doing thinks. If you have a better idea, I would be grateful :pray:

File: app/server/lib/OIDCConfig.ts
Comment (fflorent) [reply app/server/lib/OIDCConfig.ts]: Yeah, I made the message more generic.

File: app/server/lib/OIDCConfig.ts
Comment (fflorent) [reply app/server/lib/OIDCConfig.ts:45]: done!

File: app/server/lib/OIDCConfig.ts
Comment (fflorent) [reply app/server/lib/OIDCConfig.ts]: I proposed something: https://github.com/gristlabs/grist-core/pull/717/files#diff-ee4d5464063d6cfc9dc24b7b63d8858ff7b978dbe767ad2a712c93fe52f7ca1dR16

File: app/server/lib/OIDCConfig.ts
Comment (dsagal) [reply app/server/lib/OIDCConfig.ts:29]: Yes, I think so. Maybe change then the introductory sentence, e.g. "For production, use https for all URLs. For development of this module on localhost, these settings should work:"

File: app/server/lib/OIDCConfig.ts
Comment (dsagal) [reply app/server/lib/OIDCConfig.ts:96]: PKCE does seem recommended for security, so probably best to keep that.

Looking again at the issue with the cookie being available, it seems that the reason we had to switch to using a `state` variable with SAML was that the redirect could be a POST, and SameSite=Lax cookies would not be included. With OIDC, the redirect is a GET, and such cookies ARE included. So actually, we should have access to the session.

Basically, this means we should be able to store the `code_verifier` in the session, and may not need the Permit method at all. The benefit would be to remove the somewhat arbitrary 20-minute timeout, and a bit of simplification. It would be good for @paulfitz to check -- maybe there are reasons to prefer the Permit-based approach.

If something isn't working with sessions, it may be for an unrelated reason. I forget the details, but sometimes we've needed to include a call to `forceSessionChange()` to get session updates to get saved.

Comment (fflorent) [reply app/server/lib/OIDCConfig.ts:101]: > I hadn't (and haven't) looked at exactly what is being stored here, and whether it needs to be kept opaque across some round trip through another service

The `codeVerifier` is some random values that is SHA256-hashed and passed during the login phase as the `codeChallenge`, and that should be stored so it can be passed in plain text during the callback phase. The authentication server then compare whether `codeChallenge === sha256(codeVerifier)`.

I hope that's a clear explanation. If not, I am a bit in a hurry, but I can provide provide more information tomorrow.

File: app/server/lib/OIDCConfig.ts
Comment (fflorent) [reply app/server/lib/OIDCConfig.ts:32]: I took your proposal and customized it a bit.

File: app/server/lib/OIDCConfig.ts
Comment (dsagal) [inline app/server/lib/OIDCConfig.ts:68]: There is a handy `censor: true` setting you can pass, to ensure the value doesn't get logged as plaintext during server startup.

File: app/server/lib/OIDCConfig.ts
Comment (dsagal) [inline app/server/lib/OIDCConfig.ts:143]: This line probably intended while debugging, and not to commit? There are a few other console.log calls sprinkled in the code.

File: app/server/lib/OIDCConfig.ts
Comment (dsagal) [inline app/server/lib/OIDCConfig.ts:171]: Maybe would be useful to mention in top comments that this is the variable required to turn on this login system. And mention also that GRIST_OIDC_SP_HOST is optional.

**Review by dsagal — COMMENTED** [review]
Looks good to me, only some trivial comments left. Thank you! Because this change relates to security, we'll have a second reviewer for it too (our general approach to changes that may affect security).

**Comment by fflorent** [comment]
Thanks @dsagal, I took into account your remarks! :pray:

File: app/server/lib/Permit.ts
Comment (georgegevoian) [inline app/server/lib/Permit.ts:35]: I think it's ok to remove this now, right?
```suggestion
```

Comment (fflorent) [reply app/server/lib/Permit.ts]: Good catch, thanks @georgegevoian!

**Review by georgegevoian — APPROVED** [review]
Thanks @fflorent! Will wait for Dmitry's approval before merging.

**Review by dsagal — APPROVED** [review]
Thank you! One reminder: to write documentation for this new feature in https://github.com/gristlabs/grist-help (similar to help/install/saml.md), and to update grist-core README (could link to that new documentation). I think it'll be a popular configuration!

**Comment by georgegevoian** [comment]
CI failure looks unrelated. Will kick off another run in a moment.

**Pull request merged**
